### PR TITLE
Remove duplicated logic from period_helper

### PR DESCRIPTION
**Pull request by jbrockmendel**
Following on the heels of #19534, this removes a bunch of duplicate logic from period_helper and uses the implementations in np_datetime and ccalendar instead.  A handful of functions are moved directly into period, but for the most part it's deletion.

- [ ] closes #xxxx
- [ ] tests added / passed
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/_libs/src/period_helper.c [hunk 1]
```diff
@@ -373,28 +217,26 @@ static npy_int64 DtoB_weekday(npy_int64 absdate) {
...
 
-static npy_int64 DtoB_WeekendToMonday(npy_int64 absdate, int day_of_week) {
-    if (day_of_week > 4) {
-        // change to Monday after weekend
-        absdate += (7 - day_of_week);
-    }
-    return DtoB_weekday(absdate);
-}
+static npy_int64 DtoB(struct date_info *dinfo, int roll_back) {
+    int day_of_week = dayofweek(dinfo->year, dinfo->month, dinfo->day);
```
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/_libs/src/period_helper.c]: `DtoB` actually stays in place at least through the next two PRs in this sequence.  This PR is getting `dayofweek` from the implementation in src/datetime/np_datetime, so did remove the duplicated routine in period_helper.

File: pandas/_libs/src/period_helper.h [hunk 2]
```diff
@@ -155,22 +142,8 @@ int get_date_info(npy_int64 ordinal, int freq, struct date_info *dinfo);
...
-int pday_of_year(npy_int64 ordinal, int freq);
-int pweek(npy_int64 ordinal, int freq);
-int phour(npy_int64 ordinal, int freq);
-int pminute(npy_int64 ordinal, int freq);
-int psecond(npy_int64 ordinal, int freq);
-int pdays_in_month(npy_int64 ordinal, int freq);
-
-char *c_strftime(struct date_info *dinfo, char *fmt);
 int get_yq(npy_int64 ordinal, int freq, int *quarter, int *year);
+int _quarter_year(npy_int64 ordinal, int freq, int *year, int *quarter);
```
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/_libs/src/period_helper.h]: Yah, _quarter_year is actually really redundant with get_yq, so they will be merged in the PR after next.

File: pandas/_libs/tslibs/period.pyx [hunk 3]
```diff
@@ -366,19 +380,105 @@ cdef object _period_strftime(int64_t value, int freq, object fmt):
...
     return result
 
+
+# ----------------------------------------------------------------------
 # period accessors
 
 ctypedef int (*accessor)(int64_t ordinal, int freq) except INT32_MIN
 
 
+cdef int pyear(int64_t ordinal, int freq):
```
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/_libs/tslibs/period.pyx]: get_accessor_func returns pointers to these functions.  Does that affect inline-ability?

[the remark this answers is not in the document]
Comment (jreback) [reply pandas/_libs/tslibs/period.pyx]: no idea. ok can you run a perf comparison on period stuff, maybe even vs 0.21.0 so we can get a good sense if anything has changed much.

[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/_libs/tslibs/period.pyx]: The compiler didn't complain about the inlining.  Looking at timeit results the difference is indistinguishable.  Will get asv going in a few minutes.

**Comment by jbrockmendel** [comment]
No change against master (running against ~0.21.0 now)

```
asv continuous -f 1.1 -E virtualenv master HEAD -b period
[...]
    before     after       ratio
  [93c86aa1] [9e88f869]
+   60.09μs    68.30μs      1.14  period.PeriodUnaryMethods.time_to_timestamp('M')
+   16.56ms    18.38ms      1.11  period.DataFramePeriodColumn.time_setitem_period_column
-   19.84μs    17.81μs      0.90  period.PeriodUnaryMethods.time_asfreq('M')
-  971.81ns   868.77ns      0.89  period.PeriodProperties.time_property('M', 'daysinmonth')
-   21.83μs    18.90μs      0.87  period.PeriodUnaryMethods.time_now('M')

  [93c86aa1] [9e88f869]
+  851.04ns   968.69ns      1.14  period.PeriodProperties.time_property('M', 'quarter')
-    1.03μs   874.42ns      0.85  period.PeriodProperties.time_property('M', 'dayofyear')

  [93c86aa1] [9e88f869]
+  110.80μs   166.18μs      1.50  period.PeriodIndexConstructor.time_from_date_range('D')

  [93c86aa1] [9e88f869]
+  151.88μs   244.22μs      1.61  period.Indexing.time_intersection
-    1.01μs   842.80ns      0.83  period.PeriodProperties.time_property('min', 'year')
```

**Comment by jreback** [comment]
rebase as we merged the get_day function. lgtm otherwise.

**Comment by jbrockmendel** [comment]
Once this goes in I can rebase #19550, then after that we'll get rid of date_info and just use pandas_datetimestruct directly.

File: pandas/_libs/tslibs/period.pyx [hunk 4]
```diff
@@ -85,28 +84,43 @@ cdef extern from "period_helper.h":
...
     int get_yq(int64_t ordinal, int freq, int *quarter, int *year)
+    int _quarter_year(int64_t ordinal, int freq, int *year, int *quarter)
+
 
 initialize_daytime_conversion_factor_matrix()
 
+
+@cython.cdivision
+cdef char* c_strftime(date_info *dinfo, char *fmt):
+    """
```
Comment (jreback) [inline pandas/_libs/tslibs/period.pyx]: as usual want to write doc-strings for things

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks!

**Pull request merged**